Added agent demo (#37)
risingwavelabs/awesome-stream-processing: 05-agent-demo/risingwave-agent.mdx

---
title: "Use Agents to analyze data ingested into RisingWave"
---

## Overview

> Viewing data in RisingWave can be difficult and disorganized at times. A powerful tool to level the playing field is to implement an AI Agent to receive quick and accurate information about data.

Reviewer: ```suggestion
```

> Through Model Context Protocol's (MCP's) You can now easily connect any AI Agent and analyzing some data using natural language only.

Reviewer: ```suggestion
```

> In this tutorial, you will learn how to implement a custom Anthropic Agent and connect it to RisingWave servers for seamless data analysis.

Reviewer: ```suggestion
In this tutorial, you will learn how to implement a custom Anthropic Agent, integrate it with RisingWave MCP, to do simple stream processing with RisingWave.
```

## Prerequisites

* Install and run RisingWave. For detailed instructions on how to quickly get started, see the [Quick start](/get-started/quickstart/) guide.

> * Ensure you have an [Anthropic](https://console.anthropic.com/settings/keys) API key. Do not worry about tokens, this demo shouldn't spend more than -----CHECK BEFORE PUBLISHING-----.

Reviewer: ```suggestion
* Ensure you have an [Anthropic](https://console.anthropic.com/settings/keys) API key.
```

```python
class MessageHistory:
    """Manages chat history with token tracking and context management."""

    def __init__(
        self,
        model: str,
        system: str,
        context_window_tokens: int,
        client: Any,
        enable_caching: bool = True,
    ):
        self.model = model
        self.system = system
        self.context_window_tokens = context_window_tokens
        self.messages: list[dict[str, Any]] = []
        self.total_tokens = 0
        self.enable_caching = enable_caching
        self.message_tokens: list[tuple[int, int]] = (
            []
        )  # List of (input_tokens, output_tokens) tuples
        self.client = client

        # set initial total tokens to system prompt
        try:
            system_token = (
                self.client.messages.count_tokens(
                    model=self.model,
                    system=self.system,
                    messages=[{"role": "user", "content": "test"}],
                ).input_tokens

>                 - 1

Reviewer: What's this `- 1` for. Seems like a[ magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)).

            )

        except Exception:

>             system_token = len(self.system) / 4

Reviewer: As above

        while (
            self.message_tokens
            and len(self.messages) >= 2
            and self.total_tokens > self.context_window_tokens
        ):
            remove_message_pair()

            if self.messages and self.message_tokens:
                original_input_tokens, original_output_tokens = (
                    self.message_tokens[0]
                )
                self.messages[0] = TRUNCATION_MESSAGE
                self.message_tokens[0] = (
                    TRUNCATION_NOTICE_TOKENS,
                    original_output_tokens,
                )
                self.total_tokens += (
                    TRUNCATION_NOTICE_TOKENS - original_input_tokens
                )

    def format_for_api(self) -> list[dict[str, Any]]:
        """Format messages for Claude API with optional caching."""
        result = [
            {"role": m["role"], "content": m["content"]} for m in self.messages
        ]

        if self.enable_caching and self.messages:
            result[-1]["content"] = [
                {**block, "cache_control": {"type": "ephemeral"}}
                for block in self.messages[-1]["content"]
            ]
        return result
```

> The create_mcp_connection function serves as a factory to create either a stdio or sse MCP connection based on the configuration provided. The setup_mcp_connections async function loops through a list of server configs, creates connections, retrieves each server’s available tools, and dynamically generates async wrapper functions that the agent can later call. It collects both the tool definitions and their corresponding function wrappers into dictionaries for use in the agent loop

Reviewer: ```suggestion
The create_mcp_connection function serves as a factory to create either a stdio or sse MCP connection based on the configuration provided. The setup_mcp_connections async function loops through a list of server configs, creates connections, retrieves each server’s available tools, and dynamically generates async wrapper functions that the agent can later call. It collects both the tool definitions and their corresponding function wrappers into dictionaries for use in the agent loop.
```

```python

> class MCPConnection(ABC):

Reviewer: I'm not sure if we need this abstraction. Checkout https://gofastmcp.com/clients/client#creating-a-client

RISINGWAVE_HOST = 0.0.0.0
RISINGWAVE_USER = root
RISINGWAVE_PASSWORD = root
RISINGWAVE_PORT = 4566
RISINGWAVE_DATABASE = dev
RISINGWAVE_SSLMODE = disable
RISINGWAVE_TIMEOUT = 30
```

Follow the [tutorial](https://docs.risingwave.com/ingestion/advanced/generate-test-data) to create a load generator or use this:

```psql
CREATE TABLE users (
  id int,
  risk int,
  cost int,
  time timestamp
)
WITH (
  connector = 'datagen',

  fields.id.length = '1',
  fields.id.kind = 'sequence',
  fields.id.start = '1000',

  fields.risk.length = '1',
  fields.risk.kind = 'random',
  fields.risk.min = '0',
  fields.risk.max = '10',
  fields.risk.seed = '5',

  fields.cost.length = '1',
  fields.cost.kind = 'random',
  fields.cost.min = '1',
  fields.cost.max = '10000',
  fields.cost.seed = '8',

  fields.time.kind = 'random',
  fields.time.max_past = '5h',
  fields.time.max_past_mode = 'relative',

  datagen.rows.per.second = '10'
)
FORMAT PLAIN ENCODE JSON;
```

Now that you are connected to the RisingWave agent, Let's do some basic data analysis with the example above.

1. Give me the database version

2. Show me the uders table structure

3. Track the highest risk scores

> 4. Sort that mv by highest spending and display

Reviewer: Please give some sample outputs, to show what the responses look like
